File: qovis_frontend/src/js/timeline/SketchGraph.js

import {Node, Edge, Graph} from "@/js/common/Graph"
import {SketchEdgeSimpleMeta, SketchGraphSimpleMeta, SketchNodeSimpleMeta} from "@/js/timeline/SimpleMeta"

/**
 * @extends {Node<SketchNode, SketchEdge>}
 */
export class SketchNode extends Node {
  /** @type {string} */
  type = 'unknown'
  /** @type {PlanNode[][]} */
  structureSeq = []
  /** @type {boolean[]} */
  effList = []

  /** @type {SketchNodeSimpleMeta} */
  simpleMeta

  createSimpleMeta() {
    const meta = new SketchNodeSimpleMeta()
    meta.init(this)
    this.simpleMeta = meta
    return meta
  }
}

/**
 * @extends {Edge<SketchNode, SketchEdge>}
 */
export class SketchEdge extends Edge {
  /** @type {SketchEdgeSimpleMeta} */
  simpleMeta

  createSimpleMeta() {
    const meta = new SketchEdgeSimpleMeta()
    meta.init(this)
    this.simpleMeta = meta
    return meta
  }
}

/**
 * @extends {Graph<SketchNode, SketchEdge>}
 */
export class SketchGraph extends Graph {
  /** @type {int} */
  seqCnt
  /** @type {SketchGraphSimpleMeta} */
  simpleMeta

  createSimpleMeta(plans) {
    const meta = new SketchGraphSimpleMeta()
    meta.init(this, plans)
    this.simpleMeta = meta
    return meta
  }

  /**
   * @param obj
   * @param {Plan[]} plans
   * @returns {SketchGraph}
   */
  static load(obj, plans) {
    const sg = new SketchGraph()

    sg.seqCnt = obj.seqCnt

    for (const v of obj.nodes) {
      const node = sg.addNodeWithVid(new SketchNode(), v.vid)

      node.type = v.type
      node.structureSeq = v.structureSeq.map((s, i) => {
        const plan = plans[i]
        return s.map(planNodeId => plan.nodeMap.get(planNodeId))
      })
      node.effList = v.effList
    }

    sg.updateIdCounter()

    let maxVid = -1
    for (const node of sg.nodeMap.values()) {
      maxVid = Math.max(maxVid, node.vid)
    }
    sg.vidCounter.set(maxVid + 1)

    for (const e of obj.edges) {
      const src = sg.nodeMap.get(e.src)
      const dst = sg.nodeMap.get(e.dst)
      sg.addEdge(new SketchEdge(), src, dst)
    }

    return sg
  }
}

File: qovis_backend/common/sketch_graph.py
from typing import List, Dict

from common.plan import PlanNode
from utils.id_counter import IdCounter


class SNode:
    def __init__(self, vid: int):
        self.vid: int = vid
        self.in_edges: List[SEdge] = []
        self.out_edges: List[SEdge] = []
        self.type: str = 'unknown'
        self.structure_seq: List[List[PlanNode]] = []
        self.eff_lst: [bool] = []


class SEdge:
    def __init__(self, src: SNode, dst: SNode):
        super().__init__()
        self.eid: str = SEdge.get_id(src, dst)
        self.src: SNode = src
        self.dst: SNode = dst

    @staticmethod
    def get_id(src: SNode, dst: SNode) -> str:
        return f'{src.vid}-{dst.vid}'


class SGraph:
    def __init__(self):
        self.vid_counter: IdCounter = IdCounter()
        self.node_dict: Dict[int, SNode] = {}
        self.edge_dict: Dict[str, SEdge] = {}
        self.seq_cnt = 0

    def add_node(self) -> SNode:
        node = SNode(self.vid_counter.get())
        self.node_dict[node.vid] = node
        return node

    def delete_node(self, vid: int) -> SNode:
        node = self.node_dict[vid]
        for e in node.in_edges + node.out_edges:
            self.delete_edge(e.eid)
        del self.node_dict[vid]
        return node

    def add_edge(self, src: SNode, dst: SNode) -> SEdge:
        edge = SEdge(src, dst)
        src.out_edges.append(edge)
        dst.in_edges.append(edge)
        self.edge_dict[edge.eid] = edge
        return edge

    def delete_edge(self, eid: str) -> SEdge:
        edge = self.edge_dict[eid]
        edge.src.out_edges.remove(edge)
        edge.dst.in_edges.remove(edge)
        del self.edge_dict[eid]
        return edge

    def nodes(self) -> List[SNode]:
        return list(self.node_dict.values())

    def edges(self) -> List[SEdge]:
        return list(self.edge_dict.values())

    def dump(self) -> Dict:
        nodes = []
        for v in self.node_dict.values():
            nodes.append({
                'vid': v.vid,
                'type': v.type,
                'structureSeq': [[pn.vid for pn in s] for s in v.structure_seq],
                'effList': v.eff_lst,
            })
        edges = []
        for e in self.edge_dict.values():
            edges.append({
                'src': e.src.vid,
                'dst': e.dst.vid,
            })
        return {
            'seqCnt': self.seq_cnt,
            'nodes': nodes,
            'edges': edges,
        }

